src/cts/src/HTreeBuilder.cpp

```diff
@@ -363,103 +346,463 @@ void adjustToplevelTopology(Point<double>& a,
+                                scalingFactor,
+                                x1,
+                                y1,
+                                x2,
+                                y2,
+                                bestLoc,
+                                sinkDist,
+                                bestSinkDist);
+  }
+
+  if (bestLoc != noBest) {
+    return true;
+  }
   return false;
 }
 
-void findLegalPlacement(
-    const Point<double>& pt,
-    unsigned leng,
+void HTreeBuilder::checkLegalityAndCostSpecial(
+    Point<double> oldLoc,
+    Point<double> newLoc,
+    Point<double> parentPoint,
+    double targetDist,
+    const std::vector<Point<double>>& sinks,
+    int scalingFactor,
     double x1,
     double y1,
     double x2,
     double y2,
-    std::vector<Point<double>>& points  // candidate new locations
-)
+    Point<double>& bestLoc,
+    double& sinkDist,
+    double& bestSinkDist)
 {
-  double px = pt.getX();
-  double py = pt.getY();
-  std::vector<Point<double>> temp;
-  for (int i = 0; i < 2; ++i) {
-    double x = (i == 0) ? x1 : x2;
-    double y = (i == 0) ? y1 : y2;
-    double dx = leng - abs(px - x);
-    double dy = leng - abs(py - y);
-    if (x >= px - leng && x <= px + leng) {
-      temp.emplace_back(x, py + dx);
-      temp.emplace_back(x, py - dx);
-    }
-    if (y >= py - leng && y <= py + leng) {
-      temp.emplace_back(px + dy, y);
-      temp.emplace_back(px - dy, y);
+  if (fuzzyEqual(newLoc.computeDist(parentPoint), targetDist)
+      && checkLegalitySpecial(newLoc, x1, y1, x2, y2, scalingFactor)) {
+    sinkDist = weightedDistance(newLoc, oldLoc, sinks);
+    if (sinkDist < bestSinkDist) {
+      bestLoc = newLoc;
+      bestSinkDist = sinkDist;
     }
+    // clang-format off
+    debugPrint(logger_, CTS, "legalizer", 3, "adjustBestLegalLoc: branchPt move:{}=>{} is legal, "
+               "dist={:0.3f}, sinkDist={:0.3f}", oldLoc, newLoc, targetDist, sinkDist);
+    // clang-format on
+  }
+  // clang-format off
+  debugPrint(logger_, CTS, "legalizer", 3, "adjustBestLegalLoc: branchPt move:{}=>{} is illegal or dist "
+	     "{:0.3f} != {:0.3f}", oldLoc, newLoc, newLoc.computeDist(parentPoint), targetDist);
+  // clang-format off
+}
+
+// 1) Branch point couldn't be legalized by simply moving it along blockage
+//    boundary, or
+// 2) Branch point is legal but parent point has moved, so it is necessary to
+//    move it to match target topology length
+// In both cases, move branch point to match target distance and minimize
+// weighted sink distance
+Point<double> HTreeBuilder::adjustBeyondBlockage(
+    Point<double> branchPoint,
+    Point<double> parentPoint,
+    double targetDist,
+    const std::vector<Point<double>>& sinks,
+    int scalingFactor)
+{
+  double px = parentPoint.getX();
+  double py = parentPoint.getY();
+  std::vector<Point<double>> candidates;
+  Point<double> point = branchPoint;
+
+  // try points that are on edges of "Manhattan square"
```

Comment:
Isn't this a Manhattan circle?

Reply:
No, actually this is a square because of the way distance is computed.